packages/fetchai/connections/ledger_api/base.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains base classes for the ledger API connection."""
+import asyncio
+from abc import ABC, abstractmethod
+from asyncio import Task
+from concurrent.futures._base import Executor
+from typing import Any, Callable, Optional
+
+from aea.configurations.base import PublicId
+from aea.crypto.registries import Registry
+from aea.helpers.dialogue.base import Dialogues
+from aea.mail.base import Envelope
+from aea.protocols.base import Message
+
+
+CONNECTION_ID = PublicId.from_str("fetchai/ledger_api:0.1.0")
+
+
+class RequestDispatcher(ABC):
+    """Base class for a request dispatcher."""
+
+    TIMEOUT = 3
+    MAX_ATTEMPTS = 120
+
+    def __init__(
+        self,
+        loop: Optional[asyncio.AbstractEventLoop] = None,
+        executor: Optional[Executor] = None,
+    ):
+        """
+        Initialize the request dispatcher.
+
+        :param loop: the asyncio loop.
+        :param executor: an executor.
+        """
+        self.loop = loop if loop is not None else asyncio.get_event_loop()
+        self.executor = executor
+
+    async def run_async(self, func: Callable[[Any], Task], *args):
+        """
+        Run a function in executor.
+
+        :param func: the function to execute.
+        :param args: the arguments to pass to the function.
+        :return: the return value of the function.
+        """
+        try:
+            response = await self.loop.run_in_executor(self.executor, func, *args)
+            return response
+        except Exception as e:
+            return self.get_error_message(e, *args)
+
+    def dispatch(self, envelope: Envelope) -> Task:
```

Comment:
this is the main method of this class. It calls the abstract method to change behaviour depending on the concrete use case.